# SPDX-License-Identifier: Apache-2.0
# Copyright (c) 2025 NVIDIA Corporation

import argparse
import functools
import logging
from concurrent import futures

from alpasim_grpc.v0.common_pb2 import AvailableScenesReturn, Empty, VersionId
from alpasim_grpc.v0.physics_pb2 import (
    PhysicsGroundIntersectionRequest,
    PhysicsGroundIntersectionReturn,
)
from alpasim_grpc.v0.physics_pb2_grpc import (
    PhysicsServiceServicer,
    add_PhysicsServiceServicer_to_server,
)
from alpasim_physics import VERSION_MESSAGE
from alpasim_physics.backend import PhysicsBackend
from alpasim_physics.utils import (
    aabb_to_ndarray,
    pose_grpc_to_ndarray,
    pose_status_to_grpc,
)
from alpasim_utils.artifact import Artifact

import grpc

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s.%(msecs)03d %(levelname)s:\t%(message)s",
    datefmt="%H:%M:%S",
)
logger = logging.getLogger(__name__)


class PhysicsSimService(PhysicsServiceServicer):
    def __init__(
        self,
        server: grpc.Server,
        artifact_glob: str,
        cache_size: int = 2,
        use_ground_mesh: bool = False,
        visualize: bool = False,
    ) -> None:
        self.server = server
        self.artifacts = Artifact.discover_from_glob(
            artifact_glob, use_ground_mesh=use_ground_mesh
        )

        self.visualize = visualize
        logger.info(f"Available scenes: {list(self.artifacts.keys())}.")

        # instantiate the method here to avoid caching `self`
        @functools.lru_cache(maxsize=cache_size)
        def get_backend(scene_id: str) -> PhysicsBackend:
            if scene_id not in self.artifacts:
                raise KeyError(f"Scene {scene_id=} not available.")

            artifact = self.artifacts[scene_id]
            logger.info(f"Cache miss, loading {artifact.scene_id=}")
            mesh_ply = artifact.mesh_ply

            # don't keep the .ply file once the backend is evicted from the lru_cache
            artifact.clear_cache()

            return PhysicsBackend(
                mesh_ply,
                visualize=self.visualize,
            )

        self.get_backend = get_backend

    def ground_intersection(
        self, request: PhysicsGroundIntersectionRequest, context: grpc.ServicerContext
    ) -> PhysicsGroundIntersectionReturn:
        logger.info(f"Recieved request for scene_id={request.scene_id}")
        logger.debug(f"full request={request}")
        try:
            backend = self.get_backend(request.scene_id)

            other_updates = []
            for other in request.other_objects:
                other_updates.append(
                    backend.update_pose(
                        pose_grpc_to_ndarray(other.pose_pair.future_pose),
                        aabb_to_ndarray(other.aabb),
                        request.future_us,
                    )
                )

            if request.HasField("ego_data"):
                # if ego data is provided
                predicted_pose = pose_grpc_to_ndarray(
                    request.ego_data.pose_pair.future_pose
                )
                ego_aabb = aabb_to_ndarray(request.ego_data.aabb)

                updated_ego_pose, updated_ego_status = backend.update_pose(
                    predicted_pose, ego_aabb, request.future_us
                )

                response = PhysicsGroundIntersectionReturn(
                    ego_pose=pose_status_to_grpc(updated_ego_pose, updated_ego_status),
                    other_poses=[
                        pose_status_to_grpc(pose, status)
                        for pose, status in other_updates
                    ],
                )
            else:
                response = PhysicsGroundIntersectionReturn(
                    other_poses=[
                        pose_status_to_grpc(pose, status)
                        for pose, status in other_updates
                    ],
                )
            logger.info("sending response")
            return response
        except Exception as e:
            context.set_code(grpc.StatusCode.UNKNOWN)
            context.set_details(str(e))
            raise

    def get_version(self, request: Empty, context: grpc.ServicerContext) -> VersionId:
        logger.info("get_version")
        try:
            return VERSION_MESSAGE
        except Exception as e:
            context.set_code(grpc.StatusCode.UNKNOWN)
            context.set_details(str(e))
            raise

    def get_available_scenes(
        self, request: Empty, context: grpc.ServicerContext
    ) -> AvailableScenesReturn:
        logger.info("get_available_scenes")
        return AvailableScenesReturn(scene_ids=list(self.artifacts.keys()))

    def shut_down(self, request: Empty, context: grpc.ServicerContext) -> Empty:
        logger.info("shut_down")
        context.add_callback(self._shut_down)
        return Empty()

    def _shut_down(self) -> None:
        self.server.stop(0)


def parse_args(
    arg_list: list[str] | None = None,
) -> tuple[argparse.Namespace, list[str]]:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--artifact-glob",
        type=str,
        help="Glob expression to find artifacts. Must end in .usdz to find relevant files.",
        required=True,
    )
    parser.add_argument("--host", type=str, default="localhost")
    parser.add_argument("--port", type=int, default=8080)
    parser.add_argument("--use-ground-mesh", type=bool, default=False)
    parser.add_argument("--visualize", type=bool, default=False)
    args, overrides = parser.parse_known_args(arg_list)
    return args, overrides


def main(arg_list: list[str] | None = None) -> None:
    args, _ = parse_args(arg_list)
    logger.info(f"Identifying as\n{VERSION_MESSAGE}")

    server = grpc.server(futures.ThreadPoolExecutor(max_workers=1))
    address = f"{args.host}:{args.port}"
    server.add_insecure_port(address)

    service = PhysicsSimService(
        server,
        args.artifact_glob,
        use_ground_mesh=args.use_ground_mesh,
        visualize=args.visualize,
    )
    add_PhysicsServiceServicer_to_server(service, server)

    logger.info(f"Serving on {address}")
    server.start()
    server.wait_for_termination()


if __name__ == "__main__":
    main()
